### PR TITLE
Fix schema generator generating duplicate type properties for the OAuthGrant subtypes

### PR DESCRIPTION
**Pull request by Ekhorn**
## Description

This PR resolves an issue where the `OAuthGrant` Value Descriptor would generate with duplicate `grant_type` (discriminator) properties, causing the generator to include an `allOf` wrapper. The `allOf` combinator is not supported in the JSON Forms. Therefore we normalize the schema to our needs.

I also wrote an additional test to check that none of our classes generate with an `allOf` wrapper and multiple UI improvements.

## Changelog

- The main issue that the OAuthGrant wouldn't render at all is resolved by d4288f20c6104068e4236c04503cb07175327680
- Then other possible instances of `allOf` keywords appearing have been resolved in f46ee696ee004ecddc0fb97f4c9a97edb347e0a1
- I've also updated the README to reflect at least a first overview of the component in cdc157824263db3f7da74392ea0428ad1607002e
- I've reverted the following fallback 787621ada16b3e4562cd9e5f3be9ace742b5b40c, because we want users to be able to select other subtypes.
- I then added a clear button and improved how default values in JSON Schemas are handled  (prevents a possible infinite loop for booleans with a default), see 10f5b58f7406207d7e1d57408f4cf694c18c44a5.
- I saw that the OAuthGrant had properties which were required, but this wasn't reflected in the schema (resolved in 8452e2c491eba94688243be2198d531d0683d41e)
- I've also made the UI in/around the JSON Forms more responsive in https://github.com/openremote/openremote/pull/2558/changes/2c04ac60db40dd6418f27b3b14fd85072d70e9be (because the clear button would overlap with other controls)
- I've applied a refactor to the SimulatorAgentLink timezone property default handling in 0e560110db3b9e8d59199dc0097a2939fe73fc53 (after the changes of 10f5b58f7406207d7e1d57408f4cf694c18c44a5)
- I saw the WebSocketSubscription `body` property had a generated title "AnyType" rather than "Body" (resolved in 86c11178b88c0f6a3d351c827a66ee8a8350ae5e) and also for the `headers` property on the `WebsocketHTTPSubscription` (resolved in 9b1d379e30345f43e962c43b784081d9c8c2a511)

## Checklist

- [ ] 1. Acceptance criteria of the linked issue(s) are met
- [ ] 2. Tests are written and all tests pass
- [ ] 3. Changes are manually tested by you and the reviewer
- [ ] 4. Documentation is written or updated

### CODE REVIEW

File: ui/component/or-attribute-input/src/index.ts [hunk 1]
```diff
@@ -203,11 +203,26 @@ export const jsonFormsInputTemplateProvider: (fallback: ValueInputProvider) => V
                     or-loading-wrapper {
                         width: 100%;
                     }
+                    or-json-forms {
+                        max-width: 100%;
+                        min-width: 0; /**/
```
Comment (Ekhorn) [inline ui/component/or-attribute-input/src/index.ts]:
```suggestion
                        min-width: 0; /* Allows the element to shrink */
```

File: ui/component/or-components/src/or-collapsible-panel.ts [hunk 2]
```diff
@@ -73,6 +73,11 @@ const style = css`
         display: flex;
         flex-direction: row;
         overflow: hidden;
+        min-width: 0; /* TODO: explain */
```
Comment (Ekhorn) [inline ui/component/or-components/src/or-collapsible-panel.ts]:
```suggestion
        min-width: 0; /* Allows the element to shrink */
```

File: ui/component/or-components/src/or-collapsible-panel.ts [hunk 3]
```diff
@@ -73,6 +73,11 @@ const style = css`
...
+    }
+
+    #header-title {
+        min-width: 0; /* TODO: explain */
```
Comment (Ekhorn) [inline ui/component/or-components/src/or-collapsible-panel.ts]:
```suggestion
        min-width: 0; /* Allows the element to shrink */
```

File: model/src/main/java/org/openremote/model/auth/OAuthGrant.java [hunk 4]
```diff
@@ -22,6 +22,8 @@
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.fasterxml.jackson.annotation.JsonSubTypes;
 import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import jakarta.validation.constraints.NotNull;
+import org.openremote.model.util.JSONSchemaUtil;
```
Comment (Ekhorn) [inline model/src/main/java/org/openremote/model/auth/OAuthGrant.java]:
```suggestion
```

File: ui/component/or-attribute-input/src/index.ts [hunk 5]
```diff
@@ -203,11 +203,26 @@ export const jsonFormsInputTemplateProvider: (fallback: ValueInputProvider) => V
...
                 <or-loading-wrapper ${ref(loadingWrapper)} .loading="${true}">
+                    <div style="display: flex; width: 100%">
                     <or-json-forms .renderers="${jsonFormsAttributeRenderers}" ${ref(jsonForms)}
                                    .disabled="${disabled}" .readonly="${readonly}" .label="${label}"
-                                   .schema="${schema}" .uischema="${uiSchema}" .onChange="${onChanged}"></or-json-forms>
+                                   .schema="${schema}" .uischema="${uiSchema}" .onChange="${onChanged}">
+                    </or-json-forms>
+                    ${when(clear, () => html`
+                        <or-mwc-input id="clear" outlined .type="${InputType.BUTTON}" icon="backspace" @or-mwc-input-changed="${
+                            () => valueChangeNotifier({ value: null })
```
Comment (Ekhorn) [inline ui/component/or-attribute-input/src/index.ts:222]: Check whether it makes sense to use `null` rather than `undefined`.

**Comment by wborn** [comment]
@codex review

File: model/src/main/java/org/openremote/model/auth/OAuthGrant.java [hunk 6]
```diff
@@ -22,6 +22,7 @@
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.fasterxml.jackson.annotation.JsonSubTypes;
 import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import jakarta.validation.constraints.NotNull;
 import org.openremote.model.util.TextUtil;
```
[the remark this answers is not in the document]
Comment (Ekhorn) [reply model/src/main/java/org/openremote/model/auth/OAuthGrant.java:27]: Already addressed

File: ui/component/or-json-forms/src/layouts/layout-vertical-element.ts [hunk 7]
```diff
@@ -86,6 +83,22 @@ const style = css`
...
+        display: flex;
+        min-width: 0; /* Allows slotted element to shrink */
+        & span {
+            align-content: center;
+            overflow: hidden;
+            text-overflow: ellipsis;
+            white-space: nowrap;
+            word-break: break-all;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Ekhorn) [reply ui/component/or-json-forms/src/layouts/layout-vertical-element.ts:101]: Based on I don't see the problem here (all supported..) https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Nesting/Using

File: ui/component/or-json-forms/src/controls/control-array-element.ts [hunk 8]
```diff
@@ -77,6 +77,22 @@ const style = css`
...
+        display: flex;
+        min-width: 0; /* Allows slotted element to shrink */
+        & span {
+            align-content: center;
+            overflow: hidden;
+            text-overflow: ellipsis;
+            white-space: nowrap;
+            word-break: break-all;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Ekhorn) [reply ui/component/or-json-forms/src/controls/control-array-element.ts:95]: Based on I don't see the problem here (all supported..) https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Nesting/Using

File: ui/component/or-json-forms/README.md [hunk 9]
```diff
@@ -1,107 +1,120 @@
...
-* `click` - Standard click event is fired when the map itself is clicked
-* `OrMapMarkerEvent.CLICKED` - A map marker has been clicked; details contains the clicked `marker`
-* `OrMapMarkerEvent.CHANGED` - A map marker has been modified; details contains the changed `marker` and name of changed
-`property`
-
 ## Supported Browsers
-The last 2 versions of all modern browsers are supported, including Chrome, Safari, Opera, Firefox, Edge. In addition,
-Internet Explorer 11 is also supported.
-
+The last 2 versions of all modern browsers are supported, including Chrome, Safari, Opera, Firefox, Edge.
```
[the remark this answers is not in the document]
Comment (Ekhorn) [reply ui/component/or-json-forms/README.md:117]: Not going to bother.. will be done in a separate issue.

File: model/src/main/java/org/openremote/model/util/JSONSchemaUtil.java [hunk 10]
```diff
@@ -563,45 +543,90 @@ public void overrideTypeAttributes(ObjectNode attrs, TypeScope scope, SchemaGene
...
-        private void addDefaultToDiscriminator(ObjectNode attrs, SchemaGenerationContext context, String customTypeKey) {
+        private void normalizeDiscriminator(ObjectNode attrs, Class<?> type, SchemaGenerationContext context, JsonTypeInfo annotation) {
+            if (annotation == null) return;
+
+            // If EXTERNAL_PROPERTY specified, manually add 'const', 'default' and 'required' properties
+            if (annotation.include() == JsonTypeInfo.As.EXTERNAL_PROPERTY) {
+                attrs.withArray(context.getKeyword(SchemaKeyword.TAG_REQUIRED))
+                        .add(context.getKeyword(SchemaKeyword.TAG_TYPE));
+                attrs.withObject(context.getKeyword(SchemaKeyword.TAG_PROPERTIES))
+                        .withObject(context.getKeyword(SchemaKeyword.TAG_TYPE))
```
[the remark this answers is not in the document]
Comment (Ekhorn) [reply model/src/main/java/org/openremote/model/util/JSONSchemaUtil.java:602]: Not addressing this.

**Comment by wborn** [comment]
@codex review

File: ui/component/or-attribute-input/src/index.ts [hunk 11]
```diff
@@ -91,7 +92,7 @@ export function getAttributeInputWrapper(content: TemplateResult, value: any, lo
     }
 
     return html`
-            <div id="wrapper" class="${buttonIcon || fullWidth ? "no-padding" : "right-padding"}">
+            <div id="wrapper" style="min-width: 0; width: 100%" class="${buttonIcon || fullWidth ? "no-padding" : "right-padding"}">
```
Comment (MartinaeyNL) [inline ui/component/or-attribute-input/src/index.ts]: I'm curious how much visual impact this has.

File: ui/component/or-attribute-input/src/index.ts [hunk 12]
```diff
@@ -146,8 +147,8 @@ async function getSchema(valueDescriptor: ValueDescriptor) {
     return promise;
 }
 
-export const jsonFormsInputTemplateProvider: (fallback: ValueInputProvider) => ValueInputProviderGenerator = (fallback) => (assetDescriptor, valueHolder, valueHolderDescriptor, valueDescriptor, valueChangeNotifier, options) => {
-    if (Util.isComplexMetaItem(valueDescriptor)) {
+export const jsonFormsInputTemplateProvider: (fallback: ValueInputProvider, dialog?: boolean) => ValueInputProviderGenerator = (fallback, clear) => (assetDescriptor, valueHolder, valueHolderDescriptor, valueDescriptor, valueChangeNotifier, options) => {
```
Comment (MartinaeyNL) [inline ui/component/or-attribute-input/src/index.ts]: What does the new `dialog` parameter do? I don't see it being used somewhere in this function.

File: ui/component/or-attribute-input/src/index.ts [hunk 13]
```diff
@@ -555,7 +571,8 @@ export class OrAttributeInput extends subscribe(manager)(translate(i18next)(LitE
 
 
         // Use json forms with fallback to simple input provider
-        const valueChangeHandler = (detail: OrInputChangedEventDetail | undefined) => {
+        const valueChangeHandler = (detail: OrInputChangedEventDetail & { errors?: boolean } | undefined) => {
+            if (detail?.errors) return;
```
Comment (MartinaeyNL) [inline ui/component/or-attribute-input/src/index.ts:575]: This `errors` addition feels odd to me. Why is `valueChange` triggered with invalid values?
Or is it expected that this function is ran with unvalidated input?

File: ui/component/or-json-forms/src/controls/control-input-element.ts [hunk 14]
```diff
@@ -60,10 +60,6 @@ export class ControlInputElement extends ControlBaseElement {
             this.inputType = InputType.JSON;
         } else if (isBooleanControl(uischema, schema, context)) {
             this.inputType = InputType.CHECKBOX;
-            if (typeof this.data !== "boolean") {
-                 this.data = Boolean(this.data);
-                 this.handleChange(this.path, this.data);
-            }
```
Comment (MartinaeyNL) [inline ui/component/or-json-forms/src/controls/control-input-element.ts:66]: Why was this removed?

File: ui/component/or-json-forms/src/layouts/layout-vertical-element.ts [hunk 15]
```diff
@@ -133,7 +146,7 @@ export class LayoutVerticalElement extends LayoutBaseElement<VerticalLayout | Gr
         const header = this.minimal ? `` : html`
             <div slot="header">
                 <span>${this.label ? computeLabel(this.label, this.required, false) : ""}</span>
-                ${this.type ? html`<span id="type-label">${this.type}</span>` : ``}
+                ${this.type ? html`<span id="type-label" title="${this.type}">${this.type}</span>` : ``}
```
Comment (MartinaeyNL) [inline ui/component/or-json-forms/src/layouts/layout-vertical-element.ts:149]: Not sure what this `title=""` adds on top of the text itself being `${this.type}`

File: ui/component/or-json-forms/src/standard-renderers.ts [hunk 16]
```diff
@@ -254,25 +251,26 @@ export const anyOfOneOfControlRenderer = (state: JsonFormsStateContext, props: C
...
-        } else {
+        //     return html`
+        //         <div class="item-container no-match-container"><span>${label}:</span><b><or-translate value="validation.noSchemaMatchFound"></b><or-mwc-input .type="${InputType.BUTTON}" outlined label="json" icon="pencil" @or-mwc-input-changed="${(ev: Event) => showJson(ev)}"></or-mwc-input></div>
+        //     `;
+        // } else {
             // We have no data so show a schema picker
-            return getSchemaPicker(rootSchema, resolvedSchema, path, keyword, props.label || label, (selectedSchema => handleChange(path, selectedSchema.defaultValueCreator())))
-        }
+            return getSchemaPicker(rootSchema, resolvedSchema, props.label || label, (selectedSchema => handleChange(path, selectedSchema.defaultValueCreator())))
+        // }
```
Comment (MartinaeyNL) [inline ui/component/or-json-forms/src/standard-renderers.ts:273]: Unprocessed TODO

File: model/src/main/java/org/openremote/model/util/JSONSchemaUtil.java [hunk 17]
```diff
@@ -563,47 +553,105 @@ public void overrideTypeAttributes(ObjectNode attrs, TypeScope scope, SchemaGene
...
+
+            while (annotation == null && type.getSuperclass() != Object.class) {
+                type = type.getSuperclass();
+
+                if (type == null) return annotation;
+
+                annotation = type.getAnnotation(JsonTypeInfo.class);
+            }
+
+            return annotation;
```
Comment (MartinaeyNL) [inline model/src/main/java/org/openremote/model/util/JSONSchemaUtil.java:586]: I think we can decrease the amount of whitespace here.

File: ui/component/or-json-forms/src/layouts/layout-vertical-element.ts (hunk 15 above)
Comment (Ekhorn) [reply ui/component/or-json-forms/src/layouts/layout-vertical-element.ts:149]: A tooltip, so you can see what type is configured.

image

File: ui/component/or-json-forms/src/controls/control-input-element.ts (hunk 14 above)
Comment (Ekhorn) [reply ui/component/or-json-forms/src/controls/control-input-element.ts:66]: Because it was replaced with a better fix. This could cause an infinite loop, please see PR description

File: ui/component/or-attribute-input/src/index.ts (hunk 13 above)
Comment (Ekhorn) [reply ui/component/or-attribute-input/src/index.ts:575]: Because AJV applies default values even when that results in errors. Until that is fixed this was the simplest fix.

File: ui/component/or-attribute-input/src/index.ts (hunk 12 above)
Comment (Ekhorn) [reply ui/component/or-attribute-input/src/index.ts]: Whoops should be `errors`, unused here, but required by the type.

File: ui/component/or-attribute-input/src/index.ts (hunk 11 above)
Comment (Ekhorn) [reply ui/component/or-attribute-input/src/index.ts]: Will show ;)

File: ui/component/or-json-forms/src/standard-renderers.ts (hunk 16 above)
Comment (Ekhorn) [reply ui/component/or-json-forms/src/standard-renderers.ts:273]: Related to https://github.com/openremote/openremote/pull/2558#discussion_r3028655760

File: model/src/main/java/org/openremote/model/util/JSONSchemaUtil.java (hunk 17 above)
Comment (Ekhorn) [reply model/src/main/java/org/openremote/model/util/JSONSchemaUtil.java:586]: We could, but not doing this without suggestion.

File: ui/component/or-attribute-input/src/index.ts [hunk 18]
```diff
@@ -146,8 +147,8 @@ async function getSchema(valueDescriptor: ValueDescriptor) {
     return promise;
 }
 
-export const jsonFormsInputTemplateProvider: (fallback: ValueInputProvider) => ValueInputProviderGenerator = (fallback) => (assetDescriptor, valueHolder, valueHolderDescriptor, valueDescriptor, valueChangeNotifier, options) => {
-    if (Util.isComplexMetaItem(valueDescriptor)) {
+export const jsonFormsInputTemplateProvider: (fallback: ValueInputProvider, errors?: boolean) => ValueInputProviderGenerator = (fallback, clear) => (assetDescriptor, valueHolder, valueHolderDescriptor, valueDescriptor, valueChangeNotifier, options) => {
```
Comment (Ekhorn) [inline ui/component/or-attribute-input/src/index.ts]:
```suggestion
export const jsonFormsInputTemplateProvider: (fallback: ValueInputProvider, clear?: boolean) => ValueInputProviderGenerator = (fallback, clear) => (assetDescriptor, valueHolder, valueHolderDescriptor, valueDescriptor, valueChangeNotifier, options) => {
```

File: ui/component/or-attribute-input/src/index.ts (hunk 11 above)
Comment (Ekhorn) [inline ui/component/or-attribute-input/src/index.ts]: The `min-width` here doesn't seems to be required?

```suggestion
            <div id="wrapper" style="width: 100%" class="${buttonIcon || fullWidth ? "no-padding" : "right-padding"}">
```

File: model/src/main/java/org/openremote/model/util/JSONSchemaUtil.java [hunk 19]
```diff
@@ -563,47 +553,105 @@ public void overrideTypeAttributes(ObjectNode attrs, TypeScope scope, SchemaGene
...
-                }
+                String activeKey;
+                if (props.has(customTypeKey)) activeKey = customTypeKey;
+                else if (props.has(typeKey)) activeKey = typeKey;
+                else return;
 
-                JsonNode typeNode = propsObj.get(typeKey);
-                if (typeNode instanceof ObjectNode typeProp) {
+                String requiredKey = context.getKeyword(SchemaKeyword.TAG_REQUIRED);
+                ArrayNode requiredArr;
```
Comment (Ekhorn) [inline model/src/main/java/org/openremote/model/util/JSONSchemaUtil.java]: Need to revisit this again, after "addressing the AI suggestions" it broke :/. Will investigate the root cause rather than adding this crap.

Comment (Ekhorn) [reply model/src/main/java/org/openremote/model/util/JSONSchemaUtil.java]: Made a temporary fix in 974833600e64fc7ca90c244745effe68f4c58e02

Comment (Ekhorn) [reply model/src/main/java/org/openremote/model/util/JSONSchemaUtil.java]: I think this will be more permanent.

File: agent/src/main/java/org/openremote/agent/protocol/websocket/WebsocketSubscription.java [hunk 20]
```diff
@@ -30,11 +31,12 @@
 })
 @JsonTypeInfo(
     use = JsonTypeInfo.Id.NAME,
-    property = "type",
-    defaultImpl = WebsocketSubscriptionImpl.class
+    property = "type"
+//    defaultImpl = WebsocketSubscriptionImpl.class
```
Comment (Ekhorn) [inline agent/src/main/java/org/openremote/agent/protocol/websocket/WebsocketSubscription.java]:
```suggestion
```

File: ui/component/or-attribute-input/src/index.ts (hunk 5 above)
Comment (Ekhorn) [reply ui/component/or-attribute-input/src/index.ts:222]: Nope doesn't clear.
